### DISCUSSION TITLE
New Embedding Models released

### DISCUSSION THREAD

**Discussion by quelsemme** [brianpetro/obsidian-smart-connections#429] [General]
Good news for the future of Smart Connections, I'd say.

TLDR:
- two new, more powerful embedding models - **3-small** and **3-large**
- token price for **3-small** reduced by 80% compared to **ada-002** ($0.00002/1K tokens) - 62500 pages per dollar over ada's 12500p/$ 
- **3-large** embeds up to 3072 dimensions (compared to ada's 1536) and is about 40% more powerful.
- **3-large** context window is 8191 tokens, same as ada. Approximately 6000 words.

@brianpetro might be able to explain whether that means 6000 words is the upper limit per note to be fully embedded? 

**from OpenAI:**
https://openai.com/blog/new-embedding-models-and-api-updates

**A new small text embedding model**
**_text-embedding-3-small_** is our new highly efficient embedding model and provides a significant upgrade over its predecessor, the **_text-embedding-ada-002_** model released in December 2022. 

**Stronger performance**. Comparing **_text-embedding-ada-002_** to **_text-embedding-3-small_**, the average score on a commonly used benchmark for multi-language retrieval (MIRACL) has increased from 31.4% to 44.0%, while the average score on a commonly used benchmark for English tasks (MTEB) has increased from 61.0% to 62.3%.

**Reduced price**. **_text-embedding-3-small_** is also substantially more efficient than our previous generation **_text-embedding-ada-002_** model. Pricing for **_text-embedding-3-small_** has therefore been reduced by 5X compared to **_text-embedding-ada-002_**, from a price per 1k tokens of $0.0001 to $0.00002.

We are not deprecating text-embedding-ada-002, so while we recommend the newer model, customers are welcome to continue using the previous generation model.

**A new large text embedding model: text-embedding-3-large**

**_text-embedding-3-large_** is our new next generation larger embedding model and creates embeddings with up to 3072 dimensions.

**Stronger performance.** **_text-embedding-3-large_** is our new best performing model. Comparing **_text-embedding-ada-002_** to **_text-embedding-3-large_**: on MIRACL, the average score has increased from 31.4% to 54.9%, while on MTEB, the average score has increased from 61.0% to 64.6%.

**Comment by brianpetro** [comment]
@quelsemme yes! This is great. The new models are already included in Smart Connections v2.

For both the small and large models, there are two options:

1) the full-size (max dimensions, largest file size, slowest performance, better results than Ada)

2) the minimal size that beats Ada performance (minimum dimensions where the output is comparable to Ada)

While I haven't had time to experiment much with them, I'm interested to learn what others think about how they compare.

Additionally, two new local models were added, and the default local model is now 4X faster than the previous model. 

🌴

**Reply by Over-The-Edge** [reply]
Thanks, that is amazing. Can you please advise the process of switching on the new models. Is it just a case of selecting say the full size model? Or do we need to select and then process embeddings or ??

**Reply by brianpetro** [reply]
Hey @Over-The-Edge

After changing the model, click the "Refresh Notes" button which is also in the settings.

"Refresh notes" is a better option than "Force refresh" because it does not delete existing embeddings.

Similarly, when you change models, the embeddings for each model will remain, so that should allow trying different models without having to reprocess all of your notes each time you switch.

Thanks for the question 😊

🌴

**Reply by Over-The-Edge** [reply]
Thanks @brianpetro very helpful and so too is your amazing Plugin!

**Reply by Over-The-Edge** [reply]
@brianpetro Sorry changed to large embedding model, refreshed notes, updated 1087 embeddings, then ran smartchat with same prompt as before that worked and now it runs and keeps running (ie blue dots looping) and doesn't produce anything. if close obsidian restart and try again the most I get out of it is one or two words. If I use Smart Connections GPT it still works as before (perhaps because I have not yet synced updates with it). Any advice appreciated. I am running the latest version of the plugin. Have same issue with a simple prompt in smart chat

**Reply by brianpetro** [reply]
@Over-The-Edge I'm also seeing intermittent errors with the chat on some models more than others.
![Screenshot 2024-01-26 at 10 18 26 PM](https://github.com/brianpetro/obsidian-smart-connections/assets/1886014/aa33cf55-d683-48c1-a6f8-1300c4102be8)

The Smart Connections GPT works differently, so that should be unaffected.

The streaming errors seem to be coming from the OpenAI API returning different/erroneous data when streaming.

The status page says everything is OK, but it wouldn't be the first time a status page was lagging behind https://status.openai.com/

 I'm about to call it a night and will check on it in the morning since it might just be a temporary issue with their API.

There are some other unrelated bugs with the chat that I also will need to address, but that will be for another day.

🤞 hopefully, things will return to normal shortly.
🌴

**Reply by brianpetro** [reply]
@Over-The-Edge Since the OpenAI API was still returning the altered format this morning, I pushed an update (`v1.6.61` and `v.2.0.73`) that patches the issue.

**Reply by Over-The-Edge** [reply]
@brianpetro Thanks very much - all working now!

**Reply by Over-The-Edge** [reply]
@brianpetro After further tests, it seems to be unable to find my "notes" now, I have refreshed notes, and even switched back to old ADA embedding - but still get "unable to find notes" in the first part of the smart chat response. When I switch embedding models I get a pop up on the right which says "Smart Connections: No results found", Do I need to do a force refresh? Any assistance appreciated

**Reply by brianpetro** [reply]
@Over-The-Edge I think you might be experiencing this issue https://github.com/brianpetro/obsidian-smart-connections/issues/430

I'm still looking into issue. I've seen it happen in my own vault, but I haven't reliably reproduced it yet.

So if anything changes for you, let me know, it might help me figure it out quicker.

🌴

**Reply by Over-The-Edge** [reply]
@brianpetro It didn't hang for me and completed the 1039 embeddings, but that was with yesterdays version. If I reprocess embeddings there are no updates. Should I consider a force refresh?

**Reply by brianpetro** [reply]
@Over-The-Edge it might be worth trying, let me know if anything changes if you try it.

**Reply by quelsemme** [reply]
@Over-The-Edge Could you try with GPT-4 8k and GPT-3.5 16k and report back on [#430](https://github.com/brianpetro/obsidian-smart-connections/issues/430)?

**Reply by Over-The-Edge** [reply]
@brianpetro Sorry I have been out, I haven't done the force refresh as found it worked with 8k and 16k. Interesting with GPT4-128k - when it produces a response in the response it says "unable to find notes" but when I use the preview option to copy context the notes are in the context. I'll post this on 430 as well

**Reply by nomadphase** [reply]
![image](https://github.com/brianpetro/obsidian-smart-connections/assets/72380765/39a00657-094b-47f1-87a6-d43e7e9a71c7)

I have been attempting to use the new text-embedding-3-large model on my vault and it has consistently been hanging at about 1/3 of the way through the process. I've attached an image of the error message.

**Reply by brianpetro** [reply]
@nomadphase one of the updates in the latest version (`v2.0.83`) was improving the OpenAI API integration. Let me know if that fixes this for you.

🌴

**Reply by nomadphase** [reply]
Hi Brian,

That is working much better for me now, thank you very much,

Notes and Blocks embedded with Open AI Text-3 Large, and smart chat with gpt-4 8k. GPT-4 128 k is still not working (no results found)

**Reply by nomadphase** [reply]
Hello Brian,

Noting that Smart Connections is prompting me to re-embed notes and blocks this morning upon reopening Obsidian. This is despite no changes having been made to any files within the vault. Not sure if this is expected?

Cheers

**Reply by brianpetro** [reply]
@nomadphase unfortunately, it's expected.

It shouldn't happen again, since the reason is because I made major changes to the backend embedding architecture, and that should be the end of it for the foreseeable future.

Thanks for letting me know, and sorry for the inconvenience!

🌴
